### DISCUSSION TITLE
Scripting + API

### DISCUSSION THREAD

**Discussion by chrisgurney** [chrisgurney/obsidian-note-toolbar#110] [Ideas]
I've started to investigate what needs to be added to allow execution of various scripts that can be written via other plugins, and what an API for Note Toolbar might look like. 

I'd love the community's thoughts on this.

> As an aside, my goal is to hopefully have minimal dependencies in the project, and not directly integrate with other plugins, in order for Note Toolbar to remain as _nimble_ and _flexible_ as possible.

## Executing scripts

These are various ways a script could be executed, a couple of which are possible today. _What else would you like to see here?_

### _Via Commands:_  Execute scripts via commands provided by plugins
_🟢 Available now!_

As an example, you can execute Templater scripts from items:
1. Put your template code in Templater’s script folder
2. Bind the template to a command in Templater’s settings
3. Add an item that executes the command

Example: [swapToolbar script](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/79) thanks to @FeralFlora

Notes:
- This script directly access the toolbar's settings object. Having APIs available that are less prone to changes would be something that might be worth investigating.

### _Via Click Handler:_ On a toolbar button
_🟢 Available now!_

1. Add an item with type `Command` with no command entered.
2. _New_ items and toolbars now have UUIDs as of [1.9.9](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.9.9). Inspect the toolbar with DevTools to grab the contents of the `id` attribute (e.g., `34bb9d4b-ca7f-4690-823f-611a94a4c7db`).
3. In your script (e.g., via dataviewjs) called from the same note, get the element and setup the click event:
```javascript
(async () => {
    function myFunction() {
        // do stuff here
    }  
    activeDocument.getElementById('49d02a76-935a-45aa-a861-ce8f7381632a').onclick = myFunction
})();
```

Notes:
- _This does not work for items in menus._
- In the next beta/release a settings update will add UUIDs to all existing items and toolbars, and I'm considering adding an item type that acts as a placeholder and reveals the UUID for easier copying/pasting.

### _Via Listener:_ On general events, such as an "activated item" event
_🟡 In development_

A `note-toolbar:item-activated` workspace event is now being emitted when an item is activated, but:

- the item info isn't passed along yet; and
- while testing with dataviewjs, I noticed that it executes the code multiple times. Will have to test from other plugin scripts, e.g., [JS Engine](https://github.com/mProjectsCode/obsidian-js-engine-plugin) and perhaps Templater?
```javascript
this.app.workspace.on("note-toolbar:item-activated", (item) => {
    console.log("This was clicked:", item); // item = "test" right now
});
```
Notes:
- This is similar to the click handler, but will work from _all_ items (toolbars and menus), and allow your script to branch based on the item triggered.
- You'll still have to make sure the item setting doesn't do anything (e.g., a `Command` type item with no command entered).

Are there other events that might be helpful?

### _Directly:_ Item type that lets you provide a script to call 
_💡 Idea_

_As suggested by @naturecodevoid in the comments:_

- This would be a way to specify a script to call, right from an item's settings.
- Note Toolbar would probably have to be made aware of where scripts live in your vault.
- A setting would be necessary to turn on before allowing script execution.
- This capability would not be designed to allow easy output to notes (but you could still do any file manipulation you want, of course).

Benefits:
- Simple UX.
- Doesn't require code blocks in the note itself.
- Keeps the call tied to / stored with the item setting (potentially something that will be reusable/copyable in the future).

Investigate:
- [ ] Any way we could just do what [`dv.view()`](https://github.com/blacksmithgu/obsidian-dataview/blob/3c29f7cb5bb76f62b5342b88050e054a7272667f/src/api/inline-api.ts#L317) does?

## API
_💡 Ideas_

In theory, your scripts will be able to more deeply integrate with Note Toolbar by importing functions that allow you to:

- Provide a callback function that can be called when a specified item is activated from a toolbar _or_ menu item.
- Change a button's label and tooltip.
  - e.g., could be useful for daily note navigation for buttons to show yesterday's / tomorrow's dates.

In my initial tests, exporting the API added _seconds_ to plugin startup time, so I'm thinking of putting API access behind a setting, as well (...unless I'm doing something terribly wrong).

## _Please share your thoughts and ideas!_

1. What ideas or use cases would you like to see made possible?
2. How do other plugins do things in this area? What do you like about what they do? What do you not like?
3. What triggers/methods/hooks would you like to see provided?

**Comment by chrisgurney** [comment]
@FelipeRearden @FeralFlora @Moyf @naturecodevoid I'd love to hear your thoughts! 💙

**Comment by amsam0** [comment]
I'd love to just be able to execute a JavaScript file on item click. It seems like this is a rather simple feature, although it may be more complex if you don't want to rely on other plugins.

I'm sure I could make either the click handler or item activated handler work, but to be honest, they both seem like pretty convoluted solutions

**Reply by chrisgurney** [reply]
Thanks for your thoughts! 

I'm curious, how exactly would you envision this working? i.e., What steps would you like to be able to take to do this?

**Reply by amsam0** [reply]
It would just be an item type called something like "JavaScript." The user would then select a file or type code directly, similar to how the command item type has the user select a command. Imo, much less convoluted than adding a listener for button presses

**Reply by chrisgurney** [reply]
I definitely like the simplicity of the UX you're looking for. It's early days on these features for me, so I will continue to investigate what options might be available.

Would you be able to provide a simple example of a button with some code and how you execute it with Meta Bind? 

Thank you for your input!

**Reply by amsam0** [reply]
Here's an example from their [docs](https://www.moritzjung.dev/obsidian-meta-bind-plugin-docs/reference/buttonactions/inlinejs/#example):
````md
```meta-bind-button
style: primary
label: Greet the World
action:
  type: inlineJS
  code: "console.log('Hello World!');"
```
````

In addition to inline JS, they also allow executing a file:
````md
```meta-bind-button
style: primary
label: JS file
action:
  type: js
  file: "path/to/file.js"
```
````

I wasn't sure if that was exactly what you were looking for, so sorry in advance if it isn't.

**Reply by chrisgurney** [reply]
That's great, thank you! (If that's how you use it, that's really what I was seeking to understand.)

**Reply by chrisgurney** [reply]
@naturecodevoid I'm curious, do your scripts generate output for display?

The advantage of Meta Bind buttons being in code blocks is that the output can be rendered in place (I'm assuming), like with Dataviewjs scripts. From a toolbar we'd have to consider where output is sent, or (perhaps initially?) not support note output at all.

**Reply by amsam0** [reply]
Currently, none of my scripts generate output. They are just customized versions of the Daily Note commands to make my workflow smoother (I prefer to copy the previous daily note instead of using a template)

**Comment by Moyf** [comment]
Wow, I'm very happy to see this!
In fact, I wanted a similar script feature—at least for formatting dates—almost from the very beginning when I started using NTB, haha.

> At that time, I wanted to perform an operation like "open today's Daily Note," hoping to input something like `date("YYYY-MM-DD")` in the Note Toolbar, which would automatically parse into the actual name and open the note. 
> However, this requirement isn't too complex, as it can be achieved using Templater scripts or commands provided by PeriodicNote.

Most of the similar complex functions I use now are implemented with TP scripts as an intermediary layer, but I'm very glad that NTB will also launch its own script support!

---

## Let me share some functions I currently use
- Insert various TP templates (clicking the NTB button is the fastest and most comfortable method!)
- Open today's Daily Note in a new tab (as I mentioned above—since the date is dynamic, I can't directly write the file name)
- Open the current page in VSCode (I currently use the **Open vault in VSCode** plugin command to achieve this)

Expanding on the third point... if the `child process` feature could be introduced, it would break through the limitations of JS within OB and **interact with other external programs**! 
That would be amazing, haha (but the risk factor would also increase, and it might require a separate setting to enable the button).

> By the way, I'm currently using the **Shell Commands** plugin to achieve similar external program interaction functionality.

---

## Feasibility of Eval
I agree with @naturecodevoid 's opinion above; as long as JS script execution is supported, almost anything can be done!

Thinking about the recent addition of support for the `url` format, I recall the `eval` feature provided by the AdvancedURI plugin ( https://publish.obsidian.md/advanced-uri-doc/Actions/Miscellaneous ). 

You just need to pass in a link like this:
```
obsidian://advanced-uri?eval=let%20e%20%3D%20app.vault.getConfig(%22readableLineLength%22)%3Bapp.vault.setConfig(%22readableLineLength%22%2C%20!e)%3Bnew%20Notice(%60%E5%B7%B2%24%7Be%3F%22%E5%85%B3%E9%97%AD%22%3A%22%E6%89%93%E5%BC%80%22%7D%E5%8F%AF%E9%98%85%E8%AF%BB%E8%A1%8C%E5%AE%BD%60)%3Bapp.workspace.activeLeaf.rebuildView()
```

It can use `eval` to parse and execute this JS code snippet (this might be the fastest native implementation solution?).

Of course, the limitations of the link itself mean it can only execute shorter, less complex scripts (and it looks quite disastrous after URL encoding).

If NTB could support:
1. Directly inputting short code snippets  
2. Executing specific JS files 

—It would obviously be better!

### Passing Parameters
If we want to implement the above functionality, providing "passing parameters" would also be necessary.

You can refer to the functionality provided by `tp.file` in Templater scripts or the second `input` parameter in `dv.view()` — 

the former allows you to **access the note context information** (note path, selected text, etc.) when executing the command, 
while the latter allows you to **pass different parameters for multiple reuse of the same script**.

**Reply by chrisgurney** [reply]
Thanks so much for your comments! Lots to sift through!

Starting here:

> Open today's Daily Note in a new tab

How do you do this?

**Reply by Moyf** [reply]
> Thanks so much for your comments! Lots to sift through!
> 
> Starting here:
> 
> > Open today's Daily Note in a new tab
> 
> How do you do this?

Oh, just double checked, I was using **Commander** plugin to create a Macro, and this Macro firstly open a new tab, then open DailyNote (Default command from core plugin DailyNote) :
![image](https://github.com/user-attachments/assets/74b12af5-97bd-4e9d-940e-2ff014553231)

And I think the OB API could also achive this:

```js
	app.workspace.getLeaf("tab").openFile(dailyNote);
```

**Comment by Moyf** [comment]
## The "Output Form" of the Script
Another situation I can think of is—how to use the script?

There may be three (or more) ways:

**1️⃣ The simplest, direct execution**
Run the script after clicking, and all subsequent operations (like opening files, etc.) are defined by the script itself.

 2️⃣ Obtain the output result of the script, and then have **some predefined operations to handle this return variable**

Such as "open as a note," "insert text into the note," etc. 

These predefined functions are more user-friendly for non-advanced users (those not very familiar with JS), as they do not need to understand the specific implementation interface.

> Myabe some more specific little pre-defined function? Yesterday I just attempted to find a plugin to "Re-order my list in selection", but I worry it might be too specialized.

 3️⃣ Execute the script and **display it as text content**

For example, previously NTB could directly parse `{{prop}}` and display it as button text; perhaps the result after executing the script could also be used as display text.

I haven't thought of very practical specific use cases yet, but what I can think of is...
Getting today's weather through an API and displaying it as button text? (`Today is [Rainy]`)
Or counting the number of notes in a certain folder and displaying it? (`Zettelkasten (25)`)

However, these are basically just for "display" purposes, and the actual significance is not very great, haha.

> Or it could combine 2 & 3, for example: get the *latest* note in a certain folder, display the note name on the button, and clicking it would open the corresponding note? 
> This kind of "dynamic query" demand should be more practical

---

As for the API part...
As far as I know, both tp and dv have certain ways to access their functions from other places—or just from JS scripts.

Perhaps consider exposing functionalities:
Temporarily switch to a certain toolbar?
Or lock the current toolbar—so it won't change with note switching—until it is unlocked?

(It would be best if it could also be called through `obsidian://` URLs!)

---

One last thing, since NTB has become my ESSENCIAL Plugin, I hope the extended function (scripts) won't impact on its performance.. at least providing an option to enable/disable it 🤣

**Reply by Moyf** [reply]
Regarding the **3️⃣ Execute the script and display it as text content**, I was thinking of a more practical use case.

For example, I currently have a button to switch between Preview and Reading Modes.

If the button text could be generated from the script, it could be:
- When I am in Preview mode, it displays **Switch to Reading**.
- Otherwise, it displays **Switch to Editing**.

This is just an example, as it makes it easy for users to distinguish between the two modes. 
However, a toggle button like this would benefit from the ability to dynamically display text generated by the script.

**Comment by chrisgurney** [comment]
Re: how to execute JavaScript, I asked @mProjectsCode his opinion about whether he thinks I need to integrate with something like JS Engine, or just point to a script and run it.

> what ever you want to do. I like JS engine, as I made it and it provides consistent globals in it's executions

> you can either copy this `.d.ts` file, or write your own type defs for the things that you need https://github.com/mProjectsCode/obsidian-js-engine-plugin/blob/master/JsEngine.d.ts

**Comment by FelipeRearden** [comment]
> @FelipeRearden I'd love to hear your thoughts! 💙

Hello @chrisgurney  and friends !

I was reflecting about this for the past days and I gonna share my ideas… 🙏

Forgive for my lack of code knowledge, let me know if something is not clear. I don’t know the technical terms.

### 1 Run .js Scripts

Be able to run `.js` scripts from a file that is in a folder inside the vault

This is important because Templater cannot run scripts when the `view state` in different from `markdown` 

Based on this item #104  . This feature could be the perfect match for canvas and kanban plugin and other plugins that has specific view state.


### 2 Taylor made Prompts, Date Picker, Suggesters from executed from NTB

NTB API will has its own UIs to be executed form the `.js`  script or from a `templater` script.

Types:

- One line Prompt
- Yes No Prompt
- Checkbook 
- Option
- Data Picker
- Suggester/Menu

In other words: after running a command from NTB, **the NTB container itself will be TRASNFORMED into a Prompt/Date Picker where the user will interact**

**NTB will be a DYNAMIC container, changing its UI based on command that is been executed.**

Since the container is always visible in the active note UI, why not use it to same screen space.

1️⃣ Real Example: I have a command in my NTB that erases the content of the active note… 

```
const content = ""
const file = tp.file.find_tfile(tp.file.title)
await app.vault.modify(file,content)
app.commands.executeCommandById('editor:focus')
```

A YES NO PROMPT from NTB that could show a message "Are you Sure?" <kbd>YES</kbd> <kbd>No</kbd>

To let me decide if I am really want to erase this note.


2️⃣ Real Example: I have a note that lives in my Right Sidebar with some TO-DOS that I have to DO every day.

A Prompt from NTB that let me easily and fast add a text to this note without hurting my screen/workspace wold be a game changer.

<hr>

2️⃣ is in my mind for a long time. After this amazing journey since the first public beta I know that you are developer that is **attention to detail** which means that I know that you know how to do beautiful UIs that works perfectly well. NTB settings is the best settings window of all obsidian plugins. That is the reason that I know that you are the right person to do this ☀️

I have workflows for all of the types that I described above. I can elaborate on this if needed 🙏

**Reply by chrisgurney** [reply]
@FelipeRearden I hope you're well!

The beginnings of an implementation of this idea is underway in the [latest beta](https://github.com/chrisgurney/obsidian-note-toolbar/releases/tag/1.16-beta-12), with more to come!

**Comment by Moyf** [comment]
I just think of a new feature request! It seems to be related to the API or scripting... 🤔

If you and others are interested, could you consider adding a **Template** option to the toolbar?
![image](https://github.com/user-attachments/assets/92a54284-c347-4e63-8087-5cd17232b2ae)

The suggestion would utilize the Templater Plugin's folder settings:
![image](https://github.com/user-attachments/assets/c63e416a-fece-40be-af3f-8db93b00e2c9)
And after clicking the button, just apply the selected Templater.

(For now, I need to add the template file in Templater's setting to register it as a command, and then choose it in **Command** option)

---

Personally, I have a LOT of Templater scripts that I would like to use.
![image](https://github.com/user-attachments/assets/07b02a91-9d00-4f97-a813-80a1ca75fbdd)

This feature would enhance the integration with other plugins. But given that Templater is a widely used plugin (ranked fifth by download count), I believe it is worth considering !
![image](https://github.com/user-attachments/assets/116046bc-f11f-4544-88bd-69e87e8ca94d)

**Reply by chrisgurney** [reply]
I still need to experiment but I have a suspicion that even if/when I get plain script execution working, the next question will be how to interact with notes, the editor, etc. 

Both Templater and Dataview provide proxies for Obsidian's objects. Whereas I would say it's _not_ in the scope of Note Toolbar to implement these proxies.

All that said, I think you might get your wish after all.

**Reply by Moyf** [reply]
Haha, for now, I would prefer to use the Obsidian API to access the note/editor (like `app.workspace.getActiveFile();`) even when using the DV/Tp script, as it is more versatile.

Regarding NTB, I believe that most of the time when we interact with the toolbar, we intend to process the active note 🤔 
(I can't think of any exceptions at the moment.)

**Comment by chrisgurney** [comment]
Comments over on Reddit from TSPhoenix:

https://www.reddit.com/r/ObsidianMD/comments/1eymbkc/comment/ljtz5h7/

**Comment by Moyf** [comment]
I encountered a problem, but I'm not sure if it can be solved.

This morning, I wrote a TP script to manipulate the cursor, but I found that when I click a button on NTB, the cursor in the mpte disappears (loses focus), causing the script to be unable to correctly obtain the cursor position.

Is it possible for NTB to add some option to restore the focus of the previous note(which is actually the current note) after clicking?

**Reply by Moyf** [reply]
![image](https://github.com/user-attachments/assets/419d3a37-0a55-4e11-ab86-3f81fa957f48)
↑ this cursor

**Reply by chrisgurney** [reply]
@Moyf Would you be able to provide an example of the script? Even if all it does is get and display the cursor position.

**Reply by FelipeRearden** [reply]
My humble attempt to help @Moyf  and @chrisgurney 

> causing the script to be unable to correctly obtain the cursor position.

In the beggining of your templater script insert these two commands before your code…

```
app.commands.executeCommandById('editor:focus')
await new Promise(r => setTimeout(r, 100));
\\your code here
```

`app.commands.executeCommandById('editor:focus') ` will get the caret cursor back to the active note before running your templater code 

I hope it helps 🙏

<hr>

@chrisgurney this is something we talked about in the first public beta: On mobile Obsidian loves to lost the focus on active note. Many years of they created this command `app.commands.executeCommandById('editor:focus')` that “solves” this issue.

If want to target this from the NTB corner, only for `COMMAND` and `URI` types, you could execute

```
app.commands.executeCommandById('editor:focus')
await new Promise(r => setTimeout(r, 100));
```

before the command / uri defined by the user. This way, the focus will always on the active note before running the command from the NTB item.

**Reply by Moyf** [reply]
Amazing!~ That works, thank you!!!

**Comment by Moyf** [comment]
```
<%*
console.log("Save Cur File");
// obsidian://advanced-uri?vault=Obsinote&commandid=editor%253Asave-file
app.commands.executeCommandById("editor:save-file");

// 获取活动视图和编辑器
const editor = app.workspace.activeLeaf.view.editor;

// 记录当前光标位置
const currentCursor = editor.getCursor();

// 获取文档内容
const content = editor.getValue();

// 查找 "## TimeRec" 标题
const lines = content.split("\n");
const timeRecIndex = lines.findIndex(line => line.trim().includes("## TimeRec"));

if (timeRecIndex !== -1) {
	// 找到标题后，计算下方两行的位置
	const targetLine = timeRecIndex + 2;
	
	// 移动光标到目标位置
	editor.setCursor({ line: targetLine, ch: 0 });
	
	// 在此处插入你想要的内容
	// 例如: editor.replaceRange("你的内容\n", { line: targetLine, ch: 0 });
	
	// 恢复原来的光标位置
	editor.setCursor(currentCursor);
} else {
	new Notice("未找到 '## TimeRec' 标题");
}
%>
```

@chrisgurney  Here is the TP script I use, which would jump the cursor under specific Heading to refresh the Dataview block.

And the codes used to get cursor position:
```js
const editor = app.workspace.activeLeaf.view.editor;
const currentCursor = editor.getCursor();
```

**Comment by chrisgurney** [comment]
@amsam0 @Moyf @FelipeRearden Watch for the beta tomorrow. Here's a sneak peek!

https://github.com/chrisgurney/obsidian-note-toolbar/wiki/Executing-scripts

**Reply by Moyf** [reply]
WOW! Looking forward to it!

**Reply by chrisgurney** [reply]
Now available! Please review this post and docs before trying it out: https://github.com/chrisgurney/obsidian-note-toolbar/discussions/171

Let me know if you have any questions in that discussion, please! **Thanks!**

**Comment by FelipeRearden** [comment]
Hello @chrisgurney and friends !

This is comment is about a new type of UI to be considered for the new NTB version in the future.

One of the most underrated modals of Obsidian is the **Notice**

```js
new Notice("Note toolbar", 10000)
```

For users with vaults that are used for many different workflows it is very common to forget the *modus operandi* that you designed months ago for a specific workflow. 

This way, I constantly use Notices inside commands that are executed from NTB to help me to remind what I have to do with this open Prompt, Suggester that is my screen right now. 

Two REAL Examples...

🅰️ README for my Reading Books Workflow;

For every workflow that I have I have a README command in my NTB that triggers a huge **Notice** with all instructions and *keyboard hotkeys* to perform each step in this workflow;

![IMG_0031](https://github.com/user-attachments/assets/e2a3005e-1392-4373-a590-42039907c806)


🅱️ Notice as a command + Dynamic Notice

The video below is from our community friend shows how the Notice could be used not just to show important static info but as command trigger and dynamic info! This is absolutely 🤯

Watch the video, I don’t need to say anything else...

https://github.com/chrisgurney/obsidian-note-toolbar/discussions/214#discussioncomment-11954969


<hr>

The main idea is to add steroids to Notice the same way that we did for `Prompt` and `Suggester`

Follow below the superpowers.

1️⃣ Markdown, HTML and Icons syntax (Badges, Iconize);
2️⃣ Hidden Placeholder for CSS customization;
~~3️⃣ Command to close the NTB Notice inside a script;~~

Let me point each: 

### 1️⃣ Markdown, HTML and Icons syntax (Badges, Iconize);

Easy to explain: be able create beautiful Notice by rendering markdown and html flags inside the Notice;

### 2️⃣ Hidden Placeholder for CSS customization;

A paramente that will be **invisible** in the Notice UI that could help power users use CSS snippets to customize **Notice** the same way that we can do to `Prompt` and `Suggester`

### ~~3️⃣ Command to close the NTB Notice inside a script;~~

~~I know that sounds weird, let me explain...~~

~~Imagine that I am running this script:~~
```js
new Notice("INSTRUCTIONS FOR THE SUGGESTERS HERE", 10000)
const values = ["value `1`", "value `2`"];
const selected1 = await NoteToolbar.suggester(values);
tR+= "**" + selected1 + "**"
//codes continues after the suggester selection
```

![IMG_0032](https://github.com/user-attachments/assets/29d256a2-7ca9-45d7-b67f-6932914a529f)


~~As you can imagine, the Notice will be active many seconds before I close the suggester:~~
~~- Otherwise, I have to manually click in the Notice to disappear.~~

~~The idea here is to have a new command that lets me define when a I want to **close** the Notice~~

‼️ UPDATE in 5 February 2025

Since I learn that I can close an open Notice using 
```javascript
let notice = new Notice("LOREM", 30000);
await notice.noticeEl.detach();
``` 

Item 3️⃣ is SOLVED

<hr> 

This is my humble idea to bring more light to Notice 🙏

Let me know if you need more information about this ☀️

**Reply by chrisgurney** [reply]
Thanks for this @FelipeRearden . Let me have a think on this one!

**Reply by chrisgurney** [reply]
Follow-up conversation here: #218 (see [comment](https://github.com/chrisgurney/obsidian-note-toolbar/discussions/218#discussioncomment-11995094)).

**Reply by chrisgurney** [reply]
Could you explain the help text in your Notice example a bit more? What exactly do the items in the list represent to you, why are they important to reference, and what do the hot keys do?

**Reply by FelipeRearden** [reply]
> Could you explain the help text in your Notice example a bit more?

Sure!

> What exactly do the items in the list represent to you,

This Notice is a templater Script triggered from <kbd>CMD OPT /</kbd> 

This is my Book processing workflow. Since I don’t like zettelkasten I developed my own sustem on how to get/store/link/search knowledge from Books.

Since it is a complex system I need a way to **remember** what I have to do to execute this workflow

> why are they important to reference, 
> and what do the hot keys do?

This a list of every step that I need to when processing/creating a Book Notes.  Each step a number of actions and for some of these actions I have specific hotkeys that execute a specific command.

 Since I don’t remember every step I trigger this Notice many times just to remember what I have to do next.

**Comment by laktiv** [comment]
New sample/simple script. Run command pallet with command of one plugin.
> Description: Some plugins have many commands and create separates buttons as not an option. Same time, typing most common words in command pallet you will see long list of various commands and should look through the list from wanted plugin. Here is example how easier to find "format table" command.
> ![runcmd](https://github.com/user-attachments/assets/7280282a-c012-43ff-98e0-5166deaf9277)

```javascript
export async function RunCmdPlugin(engine, args) {
  const prefix = args['plugin'];

  const filteredEntries = Object.entries(app.commands.commands)
    .filter(([key]) => key.startsWith(prefix));

  let name;
  for (const [key, value] of Object.entries(app.plugins.manifests)) {
    if (value.id === prefix)
      name = value.name;
      break;
  }

  const keys = filteredEntries.map(([key]) => key);
  const values = filteredEntries.map(([, value]) => value.name);

  const selected = await ntb.suggester(values, keys, {
    placeholder: "❇️ Pick command of < "+ name +" > to tun:"
  });

  if (selected) {
    try {
      app.commands.executeCommandById(selected);
    } catch (error) {
      console.error("Command run failed:", error);
      new Notice(`Command run failed.\nSee console for info.`);
    }
  }
}
```
Running instructions:
![image](https://github.com/user-attachments/assets/730b9a57-eaa4-4eca-866b-92f71a07dd98)

Now script have not enough error handling and doesn't show assigned keystrokes on the right of suggester. Don't have enough skills. If someone will updates it, please share and tag me. Thank you.
